plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/FilesTable.java

```diff
@@ -95,82 +106,153 @@ public ConnectorTableMetadata getTableMetadata()
-            if (checkNonNull(dataFile.upperBounds(), pagesBuilder)) {
-                pagesBuilder.appendIntegerVarcharMap(dataFile.upperBounds().entrySet().stream()
-                        .filter(entry -> idToTypeMapping.containsKey(entry.getKey()))
-                        .collect(toImmutableMap(
-                                Map.Entry<Integer, ByteBuffer>::getKey,
-                                entry -> Transforms.identity().toHumanString(
-                                        idToTypeMapping.get(entry.getKey()), Conversions.fromByteBuffer(idToTypeMapping.get(entry.getKey()), entry.getValue())))));
+        PlanFilesIterable planFilesIterable = new PlanFilesIterable(tableScan.planFiles(), idToTypeMapping, types, typeOperators);
+        return planFilesIterable.cursor();
+    }
+
+    private static class PlanFilesIterable
+            extends CloseableGroup
+            implements Iterable<List<Object>>
+    {
+        private CloseableIterable<FileScanTask> planFiles;
+        private final List<io.trino.spi.type.Type> types;
+        private final TypeOperators typeOperators;
+        private final Map<Integer, Type> idToTypeMapping;
+        private CloseableIterator<FileScanTask> planFilesIterator;
+        private boolean closed;
+
+        public PlanFilesIterable(CloseableIterable<FileScanTask> planFiles, Map<Integer, Type> idToTypeMapping, List<io.trino.spi.type.Type> types, TypeOperators typeOperators)
+        {
+            this.planFiles = requireNonNull(planFiles, "planFiles is null");
+            this.types = requireNonNull(types, "types is null");
+            this.typeOperators = requireNonNull(typeOperators, "typeOperators is null");
+            this.idToTypeMapping = requireNonNull(idToTypeMapping, "idToTypeMapping is null");
+            this.planFilesIterator = planFiles.iterator();
+        }
+
+        public RecordCursor cursor()
+        {
+            return new CloseableInMemoryRecordCursor(types, this);
+        }
+
+        @Override
+        public CloseableIterator<List<Object>> iterator()
+        {
+            addCloseable(planFiles);
+            addCloseable(planFilesIterator);
+
+            return new CloseableIterator<>()
+            {
+                @Override
+                public boolean hasNext()
+                {
+                    return !closed && planFilesIterator.hasNext();
+                }
+
+                @Override
+                public List<Object> next()
+                {
+                    return getRow(planFilesIterator.next().file());
+                }
+
+                @Override
+                public void close()
+                        throws IOException
+                {
+                    PlanFilesIterable.super.close();
+                    planFiles = null;
+                    planFilesIterator = null;
+                    closed = true;
+                }
+            };
+        }
+
+        private List<Object> getRow(DataFile dataFile)
```

Comment:
`mapToRow` or `toRow` ?

Reply:
how about `getRecord`? Since we are using `RecordCursor` to iterate over.